File: sandbox/gpu_tests/bench_julia_vs_jax.jl
using CUDA
using LinearAlgebra
using BenchmarkTools
using Printf

function batch_inv_nosync!(X::CuArray{FT,3}, A::CuArray{FT,3}) where {FT}
    pivot, info = CUDA.CUBLAS.getrf_strided_batched!(A, true)
    CUDA.CUBLAS.getri_strided_batched!(A, X, pivot)
    CUDA.synchronize()
end

matrix_sizes = [12, 24, 36]
batch_sizes  = [6837, 10000, 50000]
num_iters    = 20

println("Julia CUDA batched operations benchmark")
println("GPU: ", CUDA.name(CUDA.device()))
println("="^75)

for FT in [Float64, Float32]
    println("\n── $FT ──")
    println()
    @printf("  %-8s  %8s  │ %12s  %10s  │ %12s  %10s\n",
            "size", "batch", "matmul (ms)", "μs/mat", "inv (ms)", "μs/mat")
    println("  ", "─"^71)

    for n in matrix_sizes
        for nBatch in batch_sizes
            # --- Batched multiply ---
            A = CUDA.randn(FT, n, n, nBatch)
            B = CUDA.randn(FT, n, n, nBatch)
            C = similar(A)
            CUDA.CUBLAS.gemm_strided_batched!('N', 'N', one(FT), A, B, zero(FT), C)
            CUDA.synchronize()

            mul_times = Float64[]
            for _ in 1:num_iters
                CUDA.synchronize()
                t = @elapsed begin
                    CUDA.CUBLAS.gemm_strided_batched!('N', 'N', one(FT), A, B, zero(FT), C)
                    CUDA.synchronize()
                end
                push!(mul_times, t)
            end
            mul_med = sort(mul_times)[num_iters ÷ 2] * 1000  # ms

            A = nothing; B = nothing; C = nothing
            GC.gc(); CUDA.reclaim()

            # --- Batched inversion ---
            A_cpu = randn(FT, n, n, nBatch)
            for k in 1:nBatch
                A_cpu[:,:,k] .+= FT(n) * FT.(I(n))
            end
            A_gpu = CuArray(A_cpu)
            X_gpu = similar(A_gpu)

            A_tmp = copy(A_gpu)
            batch_inv_nosync!(X_gpu, A_tmp)

            inv_times = Float64[]
            for _ in 1:num_iters
                A_tmp = copy(A_gpu)
                CUDA.synchronize()
                t = @elapsed begin
                    batch_inv_nosync!(X_gpu, A_tmp)
                end
                push!(inv_times, t)
            end
            inv_med = sort(inv_times)[num_iters ÷ 2] * 1000  # ms

            @printf("  %2d×%-4d  %8d  │ %10.3f ms  %8.3f μs  │ %10.3f ms  %8.3f μs\n",
                    n, n, nBatch,
                    mul_med, mul_med * 1000 / nBatch,
                    inv_med, inv_med * 1000 / nBatch)

            A_gpu = nothing; X_gpu = nothing
            GC.gc(); CUDA.reclaim()
        end
    end
end

File: sandbox/gpu_tests/bench_julia_vs_jax.py
import time
import os
os.environ["JAX_PLATFORMS"] = "cuda"

import jax
import jax.numpy as jnp

# Enable float64 support
jax.config.update("jax_enable_x64", True)

@jax.jit
def batched_mul(a, b):
    return jnp.matmul(a, b)

@jax.jit
def batched_inv(a):
    return jnp.linalg.inv(a)

matrix_sizes = [12, 24, 36]
batch_sizes  = [6837, 10000, 50000]
num_iters    = 20
key = jax.random.PRNGKey(0)

print(f"JAX batched operations benchmark")
print(f"JAX version: {jax.__version__}")
print(f"JAX backend: {jax.default_backend()}")
devs = jax.devices()
for d in devs:
    print(f"Device: {d}")
print("=" * 75)

for FT, label in [(jnp.float64, "Float64"), (jnp.float32, "Float32")]:
    print(f"\n── {label} ──")
    print()
    print(f"  {'size':<8s}  {'batch':>8s}  │ {'matmul (ms)':>12s}  {'μs/mat':>10s}  │ {'inv (ms)':>12s}  {'μs/mat':>10s}")
    print("  " + "─" * 71)

    for n in matrix_sizes:
        for nBatch in batch_sizes:
            A = jax.random.normal(key, (nBatch, n, n), dtype=FT)
            B = jax.random.normal(key, (nBatch, n, n), dtype=FT)

            # Warmup matmul
            C = batched_mul(A, B)
            C.block_until_ready()

            mul_times = []
            for _ in range(num_iters):
                start = time.perf_counter()
                C = batched_mul(A, B)
                C.block_until_ready()
                mul_times.append(time.perf_counter() - start)

            mul_med = sorted(mul_times)[num_iters // 2] * 1000

            # Stabilize for inversion
            eye = jnp.eye(n, dtype=FT)[None, :, :]
            A_stable = A + float(n) * eye
            A_stable.block_until_ready()

            # Warmup inv
            X = batched_inv(A_stable)
            X.block_until_ready()

            inv_times = []
            for _ in range(num_iters):
                start = time.perf_counter()
                X = batched_inv(A_stable)
                X.block_until_ready()
                inv_times.append(time.perf_counter() - start)

            inv_med = sorted(inv_times)[num_iters // 2] * 1000

            print(f"  {n:2d}×{n:<4d}  {nBatch:8d}  │ {mul_med:10.3f} ms  {mul_med*1000/nBatch:8.3f} μs  │ {inv_med:10.3f} ms  {inv_med*1000/nBatch:8.3f} μs")

            del A, B, A_stable
